spec/unit/chef_fs/config_spec.rb

```diff
@@ -103,4 +103,133 @@ def platform_path(*args)
+
+        before do
+          expect(Chef::ChefFS::PathUtils).to receive(:realest_path).with("/base_path/cookbooks", cwd).and_return("/base_path/cookbooks")
+          allow(Chef::ChefFS::PathUtils).to receive(:realest_path).with("/base_path/roles", cwd).and_return("/base_path/roles")
+        end
+
+        it "resolves a relative path" do
+          expect(Chef::ChefFS::PathUtils).to receive(:realest_path).with("blah", cwd).and_return("/base_path/cookbooks/blah")
+          expect(cfg.server_path("blah")).to eql("/cookbooks/blah")
+        end
+
+        it "resolves a relative path in a parent directory" do
+          expect(Chef::ChefFS::PathUtils).to receive(:realest_path).with("../roles/blah", cwd).and_return("/base_path/roles/blah")
+          expect(Chef::ChefFS::PathUtils).to receive(:realest_path).with("/base_path/roles", cwd).and_return("/base_path/roles")
+          expect(cfg.server_path("../roles/blah")).to eql("/roles/blah")
+        end
+
+        it "ignores a relative path that's outside the repository" do
+          expect(Chef::ChefFS::PathUtils).to receive(:realest_path).with("../../readme.txt", cwd).and_return("/readme.txt")
+          expect(cfg.server_path("../../readme.txt")).to be_nil
+        end
+
+        it "deals with splat paths" do
+          expect(Chef::ChefFS::PathUtils).to receive(:realest_path).with("*/*ab*", cwd).and_return("/base_path/cookbooks/*/*ab*")
+          expect(cfg.server_path("*/*ab*")).to eql("/cookbooks/*/*ab*")
+        end
+
+        it "resolves an absolute path" do
+          expect(Chef::ChefFS::PathUtils).to receive(:realest_path).with("/base_path/cookbooks/blah", cwd).and_return("/base_path/cookbooks/blah")
+          expect(cfg.server_path("/base_path/cookbooks/blah")).to eql("/cookbooks/blah")
+        end
+
+        it "deals with an absolute path with splats" do
+          expect(Chef::ChefFS::PathUtils).to receive(:realest_path).with("/*/cookbooks/blah", cwd).and_return("/*/cookbooks/blah")
+          expect(cfg.server_path("/*/cookbooks/blah")).to be_nil
+        end
+      end
+    end
+
+    describe "#format_path" do
+      Entry = Struct.new(:path)
+
+      let(:config) do
+        Mash.new({
+          chef_repo_path: "/base_path",
+          cookbook_path: "/base_path/cookbooks",
+          role_path: "/base_path/roles",
+        })
+      end
+
+      let (:path) { "/roles/foo.json" }
+      let (:entry) { Entry.new(path) }
+
+      it "returns the entry's path if the cwd isn't in the config" do
+        cfg = Chef::ChefFS::Config.new(config, "/my_repo/cookbooks")
+        expect(cfg).to receive(:base_path).and_return(nil)
+        expect(cfg.format_path(entry)).to eq(path)
+      end
+
+      it "returns . if the cwd is the same as the entry's path" do
+        cfg = Chef::ChefFS::Config.new(config, "/base_path/roles/foo.json")
+        expect(cfg).to receive(:base_path).and_return("/roles/foo.json").exactly(4).times
```

Comment:
I usually see `exactly` as a test smell: say it breaks and someone looks at the test, is there an intuitive way to connect the number of times `base_path` gets called with what the code is doing? (Especially if our hypothetical debugger looks at the next 2 tests and sees different numbers for `exactly`.)